Feature: Generic sensor support and data logging

mjonuschat opened the pull request:
This feature implements a sensor component that can be used to track/log generic sensors from multiple sources. Each sensor can have properties like unit of measurement, accuracy and a display name that help frontends display the tracked measurements.

Some supported sources could be locally connected I2C, SPI or 1-wire sensors as well as remote sensor providers like MQTT or REST API driven sensor.

Example of the component in use with a ESPHome device connected through MQTT:

```
[sensor.py:component_init()] - Initializing sensor component
[sensor.py:initialize()] - Registered sensor 'Chamber Temperature'
[sensor.py:initialize()] - Registered sensor 'Chamber Humidity'
[sensor.py:initialize()] - Registered sensor 'Power Usage'
[sensor.py:component_init()] - All sensors have been initialized
...
[mqtt.py:__call__()] - MQTT Subscriptions Acknowledged
 Topic: devvm-mjonuschat.sharks-with-lasers.net/moonraker/api/request | Granted QoS 0
 Topic: hank/sensor/chamber_temperature/state | Granted QoS 0
 Topic: hank/sensor/chamber_humidity/state | Granted QoS 0
 Topic: hank/sensor/pzem-004t_v3_power/state | Granted QoS 0
[sensor.py:_on_state_update()] - Context: {'payload': '27.7'}
[sensor.py:_on_state_update()] - Received updated sensor value for Chamber Temperature: 27.7°C
[sensor.py:_on_state_update()] - Context: {'payload': '44.9'}
[sensor.py:_on_state_update()] - Received updated sensor value for Chamber Humidity: 44.9%
[sensor.py:_on_state_update()] - Context: {'payload': '1.60'}
[sensor.py:_on_state_update()] - Received updated sensor value for Power Usage: 1.6W
...
```
Querying the aggregated endpoint to retrieve sensor values:
```
{
  "result": {
    "chamber_temperature": {
      "source": "hank/sensor/chamber_temperature/state",
      "name": "Chamber Temperature",
      "unit_of_measurement": "°C",
      "accuracy_decimals": 1,
      "measurements": [
        0,
        27.7,
        27.7,
        27.7,
        27.7,
        27.7,
        27.7,
        27.7,
        27.7,
        27.7,
        27.7,
        27.7,
        27.7
      ]
    },
    "chamber_humidity": {
      "source": "hank/sensor/chamber_humidity/state",
      "name": "Chamber Humidity",
      "unit_of_measurement": "%",
      "accuracy_decimals": 1,
      "measurements": [
        0,
        44.9,
        44.9,
        44.9,
        44.9,
        44.9,
        44.9,
        44.9,
        44.9,
        44.9,
        44.9,
        44.9,
        44.9
      ]
    },
    "power_usage": {
      "source": "hank/sensor/pzem-004t_v3_power/state",
      "name": "Power Usage",
      "unit_of_measurement": "W",
      "accuracy_decimals": 2,
      "measurements": [
        0,
        1.6,
        1.6,
        1.6,
        1.6,
        1.6,
        1.6,
        1.6,
        1.6,
        1.6,
        1.6,
        1.6,
        1.6
      ]
    }
  }
}
```

Signed-off-by: Morton Jonuschat <mjonuschat+moonraker@gmail.com>

Arksine commented:
Thanks.  I'm open to discussing generic sensor support, however the `aioesphomeapi` dependency will create an issue.  It depends on `noiseprotocol`, which depends on `cryptography`.  Unfortunately the Rust bindings in `cryptography` will cause a problem on SBCs that don't have prebuilt wheels available (ie: Armbian).  The wheel won't build and Moonraker will fail to update.  Initial support for ESPHome devices would probably need to be accomplished via MQTT.  In the future when Moonraker reaches stable we could revisit this, as distribution and updates for stable releases would include prebuilt dependencies.

Separately, I think this can be accomplished though a single component.  Its acceptable to break the component into separate modules if necessary, that could be accomplished by making the component a package (see the `file_manager` and `update_manager` components as examples).   I don't believe we need a separate data store module though, it seems that each base sensor could store whatever data is necessary, then `sensor.py` could expose endpoints for querying the data.

Arksine commented on moonraker/components/sensor.py:
```diff
@@ -0,0 +1,232 @@
+                "unit_of_measurement": sensor.config.unit,
+                "accuracy_decimals": sensor.config.accuracy_decimals,
+                "measurements": list(sensor.values),
+            }
```
With the above recommendations this would probably report a result something like the following:
```python
{
    "bme280": {
        "friendly_name": "bme_280",
        "values": {
            "temperature": 21.4,
            "humidity": "33"
        }
    }
}
```

Likewise, an endpoint that reports the store might look like the following:
```python
{
    "bme280": {
        "temperature": [21.4, 22.0, 21.8]
        "humidity": ["33.1, 33.1, 33.0]
    }
}
```

Arksine commented on moonraker/components/sensor.py:
```diff
@@ -0,0 +1,232 @@
+            endtime = cur_time + 120.0
+
+            uninitialized_sensors = list(self.sensors.values())
+            while cur_time < endtime:
```
This logic was taken from the `power` module before it was refactored.  There is no longer a loop in `component_init`, instead each sensor's init routine is executed asynchronously.  This init procedure is indefinite, however it does log failures.  This was done for two reasons:

1) The server won't start until `component_init` completes for each component.  We want to avoid delaying it.
2) When the machine first boots this can cause failures because the network isn't up yet.

That said, the `mqtt` implementation of initialize doesn't event need to be async.  It doesn't await anything and init occurs when it the subscription data is received.  The above will only be necessary for implementations that require polling should they be added.

Arksine commented:
Thanks, the configuration documentation provided me with a more generalized idea of what you are looking to accomplish.  I think this module adds value, and I have added some review comments.

With regard to the API, we probably need a couple of endpoints.  One that returns current device info (including its current value), and one that returns the queue stored values for each device.  I touched on what that might look like in the review comments.  The `info` endpoint should have an option for consumers to specify a specific device.

Additionally, most clients want changes in the value pushed to them over the websocket.  In the review I mentioned having a single timer that updates the queue for each registered sensor.  This timer could also push emit a `sensor_update` event that is registered as a websocket notification.  This event would only include sensors that detected a change, if no sensor values change then it wont be emitted.  A one second timer is probably appropriate.

Arksine commented on moonraker/components/sensor.py:
```diff
@@ -0,0 +1,232 @@
+        self.config = SensorConfiguration(
+            id=name,
+            name=cfg.get("name", name),
+            unit=cfg.get("unit", ""),
```
After some additional thought, I'm not sure we need a generic `fields` option either.  The template can set the fields for the value.  We could pass a method into the context called `set_result` that takes a dict and assigns the current value.  

The template could also be responsible for rounding the result, which would remove the need for the `accuracy_decimals` (or `precision`) option.

mjonuschat commented:
@Arksine : I think I've addressed most of the comments / implemented the suggested changes. Have not yet tested this fully, wanted to check in to see that this is going in the direction you were thinking.
In regards to the notifications I have implemented those with the only notify when readings changed but it feels a bit counter-intuitive for sensors where you would expect to get a time-series with predictable intervals between the readings, even if say the temperature didn't change (unless it's expected that the clients "fill the gaps").

mjonuschat commented:
Tested and verified working with the example configuration and notification have been added

Arksine commented on moonraker/components/sensor.py:
```diff
@@ -0,0 +1,315 @@
+        # Copy the last measurements data
+        self.last_value = {**self.last_measurements}
+
+        return eventtime + SENSOR_UPDATE_TIME
```
It isn't necessary to return a value here since the timer is now in the main component.

Arksine commented:
It is close, just a couple of minor issues and the "notification bundling". Additionally the APIs and websocket notification will need documentation, then it will be ready to merge.

Arksine commented on moonraker/components/sensor.py:
```diff
@@ -0,0 +1,309 @@
+    from websockets import WebRequest
+
+SENSOR_UPDATE_TIME = 1.0
+SENSOR_EVENT_NAME = ""
```
Presumably this should be `sensors:sensor_update`?

Arksine commented:
Other than the missing sensor event name the changes to the source look good.  You need to rebase your branch against master to resolve the merge conflicts with `api_changes.md`, frankly you don't need to update that if you don't want.  That is mostly for breaking changes.  Very soon I plan to keep a single changelog.

The websocket notification needs to be documented in `web_api.md`.  Since you basically have it documented in `api_changes` you could move it with a few tweaks.

mjonuschat commented:
Requested changes have been addressed and changes have been rebased. Let me know if you want me to squash the changes before merging.

Arksine commented on docs/web_api.md:
````diff
@@ -6290,6 +6438,30 @@ disconnects clients will receive a `disconnected` event with the data field
+            }
+        }
+    ]
+  ```
````
I believe the whitespace before the three backticks is causing an issue with rendering.  Also the parent json object is missing a closing `}`.

Arksine commented on docs/web_api.md:
````diff
@@ -4915,6 +4915,154 @@ State of the strip.
+```json
+{
+    "jsonrpc": "2.0",
+    "method":"machine.device_power.devices",
````
This should presumably be `server.sensors.list`.

Arksine commented on docs/web_api.md:
````diff
@@ -4915,6 +4915,154 @@ State of the strip.
+
+HTTP request:
+```http
+GET /server/sensors/list?sensor=sensor1
````
Should be `/server/sensors/sensor?sensor=sensor1`?  Looking at that, perhaps the endpoint name should be `/server/sensors/info?sensor=<sensor_name>`.  The `sensors/sensor?sensor=` is a bit much.

Arksine commented on docs/web_api.md:
````diff
@@ -4915,6 +4915,154 @@ State of the strip.
+```json
+{
+    "jsonrpc": "2.0",
+    "method": "machine.device_power.get_device",
````
Presuming we change the endpoint as recommended, this should be `server.sensor.info`.

Arksine commented on moonraker/components/sensor.py:
```diff
@@ -0,0 +1,309 @@
+        output = {
+            key: sensor.get_sensor_measurements()
+            for key, sensor in self.sensors.items()
+            if sensor_name is None or key == sensor_name
```
I may have missed this on the previous review, however `sensor_name` defaults to a null string on line 290.  This will result in an empty return value when a front end requests all measurements.

mjonuschat replied:
Sorry, missed updating the condition when mypy complained about the None as sentinel value when retrieving the request param.

Arksine commented:
Thanks.  I found some issues with the docs, and apparently I missed a problem with the `measurements` request in my last review.  I also proposed refactoring the name of `/server/sensors/sensor` to `/server/sensors/info`, as the `/sever/sensors/sensor?sensor=` is a bit extreme on the use of the term "sensor".

mjonuschat commented:
Sorry for the force-push, forgot to add the Signed-off comment to a bunch of the commit messages.

Tested the endpoint changes and the bugfix:

```
pi@moonraker-dev:~ $ curl -s "http://localhost:7125/server/sensors/measurements" | jq
{
  "result": {
    "powermeter": {
      "power": [
        0,
        0,
        0
      ],
      "voltage": [
        115.7,
        115.7,
        115.7
      ],
      "current": [
        0,
        0,
        0
      ],
      "energy": [
        0,
        0,
        0
      ]
    }
  }
}
```

```
pi@moonraker-dev:~ $ curl -s "http://localhost:7125/server/sensors/measurements?sensor=powermeter" | jq
{
  "result": {
    "powermeter": {
      "power": [
        0,
        0
      ],
      "voltage": [
        115.4,
        115.4
      ],
      "current": [
        0,
        0
      ],
      "energy": [
        0,
        0
      ]
    }
  }
}
```

```
pi@moonraker-dev:~ $ curl -s "http://localhost:7125/server/sensors/info?sensor=powermeter"|jq
{
  "result": {
    "id": "powermeter",
    "friendly_name": "powermeter",
    "type": "mqtt",
    "values": {
      "power": 0,
      "voltage": 115.4,
      "current": 0,
      "energy": 0
    }
  }
}
```

Arksine commented:
Thanks.
